src/client/serialize/daos_serialize.c

```diff
@@ -0,0 +1,941 @@
+		D_GOTO(out, rc);
+	}
+out:
+	/* close hdf5 objects */
+	if (usr_attr > 0)
+		H5Aclose(usr_attr);
+	if (attr_dspace > 0)
+		H5Sclose(attr_dspace);
+	return rc;
+}
+
+int
+daos_cont_serialize_props(hid_t file_id, daos_prop_t *prop_query)
+{
+	int			rc = 0;
+	struct daos_prop_entry	*entry;
+
+	if (prop_query == NULL) {
+		rc = -DER_NOMEM;
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[0];
+	rc = serialize_str(file_id, entry, "DAOS_PROP_CO_LABEL");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[1];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_LAYOUT_TYPE");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[2];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_LAYOUT_VER");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[3];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_CSUM");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[4];
+	rc = serialize_uint(file_id, entry,
+			    "DAOS_PROP_CO_CSUM_CHUNK_SIZE");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[5];
+	rc = serialize_uint(file_id, entry,
+			    "DAOS_PROP_CO_CSUM_SERVER_VERIFY");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[6];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_REDUN_FAC");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[7];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_REDUN_LVL");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[8];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_SNAPSHOT_MAX");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[9];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_COMPRESS");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[10];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_ENCRYPT");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[11];
+	rc = serialize_str(file_id, entry, "DAOS_PROP_CO_OWNER");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[12];
+	rc = serialize_str(file_id, entry, "DAOS_PROP_CO_OWNER_GROUP");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[13];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_DEDUP");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[14];
+	rc = serialize_uint(file_id, entry,
+				      "DAOS_PROP_CO_DEDUP_THRESHOLD");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[15];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_ALLOCED_OID");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	/* serialize ACL */
+	if (prop_query->dpp_nr > 16) {
+		entry = &prop_query->dpp_entries[16];
+		rc = serialize_acl(file_id, entry, "DAOS_PROP_CO_ACL");
+		if (rc != 0) {
+			D_GOTO(out, rc);
+		}
+	}
```

Comment:
I do see we are missing some properties here:
        DAOS_PROP_CO_ROOTS, DAOS_PROP_CO_STATUS, DAOS_PROP_CO_EC_CELL_SZ

Reply:
> I do see we are missing some properties here:
> DAOS_PROP_CO_ROOTS, DAOS_PROP_CO_STATUS, DAOS_PROP_CO_EC_CELL_SZ

yeah, fair enough. That does allow more flexibility if a different number of properties are passed into the function. I suppose this wasn't an issue previously because we always made sure we retrieved all of the properties. I will also add the other properties.

Reply:
well this is what im not sure about, is how you are using daos_cont_serialize_md()
it's probably from an external program that i don't see here. but if someone else uses it and doesn't use it as expected, this would not work. in any case, since you missed some properties, this still is wrong because for example:
dpp_entries[15]; is not DAOS_PROP_CO_ALLOCED_OID; but i think DAOS_PROP_CO_DEDUP_THRESHOLD
prop_query->dpp_entries[9]; is not DAOS_PROP_CO_COMPRESS, but it's DAOS_PROP_CO_ACL.

im jsut looking at the order in:
https://github.com/daos-stack/daos/blob/master/src/include/daos_prop.h#L95
but again im not sure how you are using this and what you pass in.
would be hard to confirm review this wihout knowing that.

Reply:
@mchaarawi yes, that is true that someone could get the properties in a different order if they used this function and it would not work. The way I use it is that for serialize I get all of the properties in the same order (and same properties) as are in the serialization. Since this is a library function I am updating it to be more flexible, where it can take properties in any order.

Reply:
@mchaarawi I wrote a function that serializes a flexible number of properties, but where it gets complicated is in the deserialization. This is because in order to read the hdf5 attribute I need to know the name of the container property (the name of the attribute in the hdf5 file), so I need to know which properties were serialized in  order to deserialize them properly. I am thinking maybe instead I can add the function I use to get the properties I serialize before calling serialize_md to make it clear that to use serialize_md you have to use serialize_get_props from this library?

Also, the DAOS_PROP_CO_ROOTS and DAOS_PROP_CO_STATUS are going to require custom serialization functions since they are structs (not just strings or uint), so I think it would be better to add support for these properties in another PR?

Reply:
@dsikich This function might be useful for flexible properties: https://github.com/daos-stack/daos/blob/0838789fc8b43ded34d17e4f25328c2803feabe6/src/common/prop.c#L548-L560

Reply:
@dsikich For deserializing properties in a flexible order, maybe we can just loop through the list of known properties and read from df5 if it exists?

Reply:
@daltonbohning not sure I understand this? When deserializing you have to read from the HDF5 file. That is how they are read into the DAOS property list. That works fine for serialization but not for deserialization. I suppose you could go through and "try" to read each property from the file, and keep going if the attribute doesn't exist.

Reply:
@dsikich Yeah, I was thinking that we could "try" to read. And if it exists, then add to the next index of dpp_entries. Otherwise, just keep going and try to read the next one. This would also be useful for backwards compatibility

Reply:
@daltonbohning ok I see what you are saying.